File: ISEbd-21-Shabatko-K.A.-laba-3-Sharp/WindowsFormsApplication4 laba3/ClassArray.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WindowsFormsApplication4
{
    class ClassArray<T> where T : ITransport
    {
        private T[] places;
        private T defaultValue;
        public ClassArray(int size, T defVal)
        {
            defaultValue = defVal;
            places = new T[size];
            for (int i = 0; i < places.Length; i++)
            {
                places[i] = defaultValue;
            }
        }
        public T getObject(int ind)
        {
            if (ind > -1 && ind < places.Length)
            {
                return places[ind];
            }
            return defaultValue;
        }
        public static int operator +(ClassArray<T> p, T plane)
        {
            for (int i = 0; i < p.places.Length; i++)
            {
                if (p.CheckFreePlace(i))
                {
                    p.places[i] = plane;
                    return i;
                }
            }
            return -1;
        }
        public static T operator -(ClassArray<T> p, int index)
        {
            if (!p.CheckFreePlace(index))
            {
                T plane = p.places[index];
                p.places[index] = p.defaultValue;
                return plane;
            }
            return p.defaultValue;
        }
        private bool CheckFreePlace(int index)
        {
            if (index < 0 || index > places.Length)
            {
                return false;
            }
            if (places[index] == null)
            {
                return true;
            }
            if (places[index].Equals(defaultValue))
            {
                return true;
            }
            return false;
        }

    }
}

File: tpL2/src/tpL2/ClassArray.java
package tpL2;

import java.util.Dictionary;
import java.util.Hashtable;

import tpL2.ClassArray;
import tpL2.ITransport;

public class ClassArray<T extends ITransport> {
	
	
	private T defaultValue;
	private Dictionary<Integer, T> places;
	int maxCount;

	public ClassArray(int size, T defVal) {
		defaultValue = defVal;
		places = new Hashtable<Integer, T>();
		maxCount = size;
	}

	public T getObject(int ind) {
		if (places.get(ind)!=null) 
			return places.get(ind);
		
		return defaultValue;
	}

	public static <T extends ITransport> int plus(ClassArray<T> p, T plane) {
		if(p.places.size()==p.maxCount) return -1;
		for (int i = 0; i < p.places.size(); i++) {
			if (p.CheckFreePlace(i)) {
				p.places.put(i, plane);
				return i;
			}
		}
		p.places.put(p.places.size(), plane);
		return p.places.size()-1;
	}

	public static <T extends ITransport> T minus(ClassArray<T> p, int ind) {
		if (p.places.get(ind)!=null) {
			T plane = p.places.get(ind);
			p.places.remove(ind);
			return plane;
		}
		return p.defaultValue;
	}

	public boolean CheckFreePlace(int index) {
		if(places.get(index)==null){
			return true;
		}

		return false;
	}

}
